kaitai/kaitaistream.cpp

```diff
@@ -425,41 +432,106 @@ std::string kaitai::kstream::bytes_terminate(std::string src, char term, bool in
 // ========================================================================
 
 std::string kaitai::kstream::process_xor_one(std::string data, uint8_t key) {
+    if (key == 0)
+        return data;
+
     size_t len = data.length();
     std::string result(len, ' ');
 
-    for (size_t i = 0; i < len; i++)
+    for (size_t i = 0; i < len; i++) {
         result[i] = data[i] ^ key;
+    }
 
     return result;
 }
 
 std::string kaitai::kstream::process_xor_many(std::string data, std::string key) {
     size_t len = data.length();
     size_t kl = key.length();
+    if (len == 1)
+        return process_xor_one(data, key[0]);
```

Comment:
Is data passed in O(1) time?

Reply:
All XOR and ROL optimisations are assuming that data is much larger than the key in xor, and group in rol. Therefore, the impl with cheaper main loop is considered better. Therefore, if data is passed in O(1) time, then yes.

Reply:
As the definition of this function stands above, the input data will be copied during the invocation.

Nevertheless, if the length of the data is 1, then it will result in only one iteration of the loop, and so the question is, which is cheaper -- executing a branch and invoking a (possibly in-lined) simple function, or setting up the loop and then exiting it after the first iteration.  I can't tell.